### ISSUE TITLE
Weight and Bias 적용해보기

### ISSUE THREAD

**Issue by ArtemisDicoTiar** [wisdomify/wisdomify#71]
![image](https://user-images.githubusercontent.com/47541511/134763006-64b757f5-88ce-4221-819f-0ad9fe2b8cea.png)

![image](https://user-images.githubusercontent.com/47541511/134762874-764dccaa-2a05-4f97-9943-bf5205e813ff.png)

![Screenshot 2021-09-25 at 20 55 01](https://user-images.githubusercontent.com/47541511/134770638-d814f45f-be93-4666-918e-4a44fe034e53.png)
계정별로 100기가의 공간을 무료로 준다!
## [Artifacts Colab](https://colab.research.google.com/github/wandb/examples/blob/master/colabs/wandb-artifacts/Pipeline_Versioning_with_W%26B_Artifacts.ipynb#scrollTo=I13myiOCyXOK)

**Comment by ArtemisDicoTiar** [comment]
우선 프라이빗으로 wandb에 wisdomify로 프로젝트를 생성했다.

**Comment by ArtemisDicoTiar** [comment]
완디비에서 제공해주는 가이드대로 1~7을 적용해보자
![Screenshot 2021-09-25 at 14 31 51](https://user-images.githubusercontent.com/47541511/134759779-d035194a-7ddf-43dd-bc11-d365fa578337.png)
.

**Comment by ArtemisDicoTiar** [comment]
# Quick Start
빠른 설정부터 ㄱㄱ

1. 파이썬 환경에서 완드비를 설치하고 로그인해야한다.
~~~bash
pip install wandb  # *주의, 만약 시스템에서 pip과 pip3을 모두 사용중이라면 pip을 써야한다. pip3이 아니다.
wandb login  # 로그인시도를 하게되면 완드비에서 api 키를 요청하는 데 시키는 대로 적혀있는 주소에 가면 바로 나온다.
~~~

2.

**Comment by ArtemisDicoTiar** [comment]
# Experiment Tracking
## W&B를 스크립트에 적용시키는 방법 4가지
1. init으로 초기화 (로그 등을 기록할 오브젝트 생성하는 듯.)
2. `wandb.config`로 하이퍼 파리미터 기록 (실험 설정)
3. `wandb.log()`로 metrics 기록 (매 training step마다, 그냥 for loop넣고 돌려도 작동한다. 즉, 반복적인 값 기록이 필요한 경우에 사용.)
    a. history 오브젝트에 매번 기록을 저장한다. (type: [{}, {}, ... ])
    b. summary 오브젝트는 매 스탭마다 history 기록과 함께 업데이트 된다.
4. `wandb.log_artifact`로 모델 wieght, 예측값 등 실행의 결과 값을 저장한다.

### Best Practice
~~~python
import wandb

# config는 트래킹할 모든 값들, reproducing에 필수적인 요소들을 넣어야한다.
config = dict (
  learning_rate = 0.01,
  momentum = 0.2,
  architecture = "CNN",
  dataset_id = "peds-0192",
  infra = "AWS",
)

# init시 작성하는 값들.
# project는 완드비 사이트에서 생성한 프로젝트 명을 적으면 됨.
# notes는 커밋 메시지 같은 거다.
wandb.init(
  project="detect-pedestrians",
  notes="tweak baseline",
  tags=["baseline", "paper1"],
  config=config,
)
~~~

![Screenshot 2021-09-25 at 15 21 32](https://user-images.githubusercontent.com/47541511/134761205-cb8e7e2e-321c-443d-a947-08186e1458c6.png)

### + α 

specific calls
* Dataset
  * 구체적으로 이미지나 데이터셋 샘플로 W&B에 스트리밍하게 해둘 수 있다.
* pytorch gradients
  * `wandb.watch(model)`
* configuration info
   * `wandb.init(config=config_dictionary)`
* metrics
   * `wandb.log(logging_dictionary)`

**Comment by ArtemisDicoTiar** [comment]
# Integration
## Pytorch-lightning Trainer
그냥 트레이너에 로거를 완드비 로거를 박으면 된다.
거기다가 라이트닝 모듈에서 완드비로거를 불러오는 거라 그냥 평소에 라이트닝에서 로그 적던 것처럼 `self.log(...)`를 사용하면 된다.
💡 주의: 라이트닝 metrics가 독립된 라이브러리로 분리되면서 (torchmetrics) 해당 기능 구현이 아직 안 된 것 같다. 
```python
from pytorch_lightning.loggers import WandbLogger  # newline 1
from pytorch_lightning import Trainer

wandb_logger = WandbLogger()  # newline 2
trainer = Trainer(logger=wandb_logger)
```

## Huggingface Trainer

```python
from transformers import TrainingArguments, Trainer

args = TrainingArguments(... , 
                         report_to="wandb",
                         run_name="bert-base-high-lr"  # name of the W&B run (optional)
)
trainer = Trainer(... , args=args)
trainer.train()

wandb.finish()
```

**Comment by ArtemisDicoTiar** [comment]
# Report
메인 대시보드 우측 상단에 보면 `create report`가 있다. 해당 버튼을 통해 레포트를 생성, 작성할 수 있다.
방법은[링크](https://docs.wandb.ai/guides/reports) 참고.

**Comment by ArtemisDicoTiar** [comment]
# Data + Model versioning
오 대충 쓱 훑어봤는데 MS 애저에 있던 파이프라이닝이랑 비슷해보인다. GUI로 직관적으로 시각화되어 있는 것도 보기 좋다. :)

**Comment by ArtemisDicoTiar** [comment]
## TODOs
- [ ] model - mlm
- [ ] model - tokenizer
- [ ] model - metrics
- [ ] datamodule

**Comment by ArtemisDicoTiar** [comment]
무작정 with 구문으로 완디비 run 오브젝트 생성 후 해당 오브젝트로 데이터 다운로드를 하니 파이프라인에 원치 않는 run이 생성된다.
그래서 생각해보니
wisdomify에서는 아래와 같은 작업만이 실행된다.

1. train
   1. data-load
  2. model-load (이 모델은 fine-tune task의 경우, 우리가 pre-train한 모델/ pre-train task의 경우, kcBERT, 그리고 연결된 tokenizer를 로드)
  3. 학습중 발생하는 로그 로깅
  4. 최종 모델 업로드 (mlm, tokenizer)
2. eval
   1. data-load
  2. model-load (이 모델은 eval 대상 모델.)
  3. evaluation 중 발생하는 로그 로깅
3. infer
   1. // data-load (유저 테스트 쿼리로 레포트 생성시 필요. 그 외는 스킵.)
  2. model-load (이 모델은 infer 대상 모델.)
  3. inference 중 발생하는 로그 로깅

그래서 이전에 TrainerFileSupport 클래스 작성한 것처럼 진행하려고한다.

완디비에 파이프라인을 올바르게 로깅하려면 반드시 run 오브젝트 하나 선언 → 데이터, 모델 사용 → 사용 → 로그 .... → run 종료
이렇게 해야하기 때문에,
1. 어떠한 클래스로 하여금 init될때 같이 wandb init을 같이 진행, 오브젝트를 클래스 내에 담게 함
2. 해당 오브젝트를 사용해서 데이터 다운로드, (데이터 업로드 → 이건 순전히 storyteller용), 모델 다운로드, 모델 업로드, 로그 로깅
3. 실행할 프로세스가 다 끝나면 wandb.finish()

**Comment by ArtemisDicoTiar** [comment]
모델 트레이닝 + 기록 아주 잘된다!!
![Screenshot 2021-10-02 at 16 19 03](https://user-images.githubusercontent.com/47541511/135707481-8294de3c-9664-43cb-999a-d5e9d7b56790.png)

**Comment by ArtemisDicoTiar** [comment]
## `conf.json`
conf.json에 최대한 많은 정보를 넣으려고한다. 코드에 하드코딩하는 건 유지보수에 굉장히 좋지 않는 방식이기 때문에...
wisdom을 리스트롤 받는 방식을 사용하는 데 이걸 w&b에 txt로 저장하고 그걸 로드하는 방식은 어떤가 생각이 든다.
일단 적어두고 나중에 변경하게되면 그때 적용.

**Comment by ArtemisDicoTiar** [comment]
@eubinecto wandb 적용 완료. 수욜날 회의 끝나고 PR review 하면서 설명해줄게.
